core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -108,8 +108,59 @@ class TransactionStateManager(brokerId: Int,
     }
   }
 
-  def enablePidExpiration() {
-    // TODO: add producer id expiration logic
+  def enableTransactionalIdExpiration() {
+    scheduler.schedule("transactionalId-expiration", () => {
+      inLock(stateLock) {
+        val now = time.milliseconds()
+
+        val transactionalIdByPartition: Map[Int, mutable.Iterable[String]] = transactionMetadataCache.flatMap { case (_, entry) =>
+          entry.metadataPerTransactionalId.filter { case (_, txnMetadata) =>
+            txnMetadata.txnLastUpdateTimestamp <= now - config.transactionalIdExpirationMs
+          }.map { case (transactionalId, _) =>
+            transactionalId
+          }
+        }.groupBy { transactionalId =>
+          partitionFor(transactionalId)
+        }
+
+        val recordsPerPartition = transactionalIdByPartition
```

Comment:
@dguy, I agree about the `addPartition` and `endTxn` cases. We should not get those requests if we are in a position to expire a transactional Id, but yet, we may. If the client suddenly starts a new transaction, it may send a valid `addPartitions` request. 

In general, doesn't there seem to be a race here? If there is no open transaction for the producer, and we decide to remove the transactional id, we could be in the process of removing it while a new transaction starts, or a new producer with the same transactionalId comes online.

To be specific, what prevents this scenario: 

0. TransactionalId `foobar` is picked for expiration.
1. An `addPartitions` with transactionalId `foobar` comes in. There is no pending state and at this point the transactionalId is in the cache, so the transaction is started. We may even update the cache with the partitions and `OnGoing` state.
2. The `append` to delete the transactionalId is written to the log.
3. the append to delete the transactionalId is written to the log, and the cache is cleared. 
4. The ongoing transaction is now corrupt. 

This may happen right? If not, what prevents it?